platforms/software/dependency-management/src/main/java/org/gradle/internal/component/ResolutionCandidateAssessor.java

```diff
@@ -0,0 +1,189 @@
+            AttributeValue<?> producerValue = candidateAttributes.findEntry(attributeName);
+
+            if (consumerValue.isPresent() && producerValue.isPresent()) {
+                if (attributeMatcher.isMatching(untyped, producerValue.coerce(attribute), consumerValue.coerce(attribute))) {
+                    compatible.add(new AssessedAttribute<>(attribute, Cast.uncheckedCast(consumerValue.get()), Cast.uncheckedCast(producerValue.get())));
+                } else {
+                    incompatible.add(new AssessedAttribute<>(attribute, Cast.uncheckedCast(consumerValue.get()), Cast.uncheckedCast(producerValue.get())));
+                }
+            } else if (consumerValue.isPresent()) {
+                onlyOnConsumer.add(new AssessedAttribute<>(attribute, Cast.uncheckedCast(consumerValue.get()), null));
+            } else if (producerValue.isPresent()) {
+                onlyOnProducer.add(new AssessedAttribute<>(attribute, null, Cast.uncheckedCast(producerValue.get())));
+            }
+        }
+    }
+
+    /**
+     * An immutable data class that holds information about a single variant which was a candidate for matching during resolution.
+     *
+     * This includes classifying its attributes into lists of compatible, incompatible, and absent attributes.  Each candidate
+     * is assessed within the context of a resolution, so this is a non-{@code static} class implicitly linked to the assessor
+     * that produced it.
+     */
+    public final class AssessedCandidate implements Describable {
+        private final String name;
+        private final ImmutableAttributes candidateAttributes;
+
+        private final ImmutableList<AssessedAttribute<?>> compatible;
+        private final ImmutableList<AssessedAttribute<?>> incompatible;
+        private final ImmutableList<AssessedAttribute<?>> onlyOnConsumer;
+        private final ImmutableList<AssessedAttribute<?>> onlyOnProducer;
+
+        private AssessedCandidate(String name, ImmutableAttributes attributes, ImmutableList<AssessedAttribute<?>> compatible, ImmutableList<AssessedAttribute<?>> incompatible, ImmutableList<AssessedAttribute<?>> onlyOnConsumer, ImmutableList<AssessedAttribute<?>> onlyOnProducer) {
+            this.name = name;
+            this.candidateAttributes = attributes;
+            this.compatible = compatible;
+            this.incompatible = incompatible;
+            this.onlyOnConsumer = onlyOnConsumer;
+            this.onlyOnProducer = onlyOnProducer;
+        }
+
+        @Override
+        public String getDisplayName() {
+            return name;
+        }
+
+        public ImmutableAttributes getAllCandidateAttributes() {
+            return candidateAttributes;
+        }
+
+        public ImmutableAttributes getAllRequestedAttributes() {
+            return requestedAttributes;
+        }
+
+        public ImmutableList<AssessedAttribute<?>> getCompatibleAttributes() {
```

Comment:
So looking at `getCompatibleAttributes`, `getOnlyOnConsumerAttributes`, and `getOnlyOnProducerAttributes`, I don't know if `ImmutableList<AssessedAttribute<?>>` is really the right type to return here. 

For compatible attributes, requested and provided are always the same. 
For only on consumer, provided will always be null
For only on producer, requested will always be null.

Does it make sense to just return an `ImmutableAttributes` for these properties?

Reply:
That's true and seems like it could work.  What's the benefit, though?

What I prefer about this way is:
- Lists of `AssessedAttribute` are a little lighter weight that `ImmutableAttributes`.
- Consistent types across all the sibling attribute collections.
- Makes it very clear to the consumer whether they are using a request or candidate value.

I don't feel very strongly about this though.  If there's a good reason to switch the types let me know.